compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs

```diff
@@ -445,13 +442,36 @@ impl<'infcx, 'tcx> MirBorrowckCtxt<'_, '_, 'infcx, 'tcx> {
                 } else {
                     (None, &[][..], 0)
                 };
+                let mut reborrow = false;
                 if let Some(def_id) = def_id
                     && let node = self.infcx.tcx.hir_node_by_def_id(def_id)
                     && let Some(fn_sig) = node.fn_sig()
                     && let Some(ident) = node.ident()
                     && let Some(pos) = args.iter().position(|arg| arg.hir_id == expr.hir_id)
                     && let Some(arg) = fn_sig.decl.inputs.get(pos + offset)
                 {
+                    // If the moved value is a mut reference, it is used in a
+                    // generic function and it's type is a generic param, it can be
+                    // reborrowed to avoid moving.
+                    // for example:
+                    // struct Y(u32);
+                    // x's type is '& mut Y' and it is used in `fn generic<T>(x: T) {}`.
+                    let is_sugg_reborrow = || {
+                        if let Some((def_id, _)) = arg.as_generic_param()
+                            && let Some(generics) = node.generics()
+                            && let Some(def_id) = def_id.as_local()
+                            && generics.params.iter().any(|param| param.def_id == def_id)
+                        {
+                            let place = &self.move_data.move_paths[mpi].place;
+                            let ty = place.ty(self.body, self.infcx.tcx).ty;
+                            if let ty::Ref(_, _, hir::Mutability::Mut) = ty.kind() {
+                                return true;
+                            }
+                        }
+                        false
+                    };
+                    reborrow = is_sugg_reborrow();
+
                     let mut span: MultiSpan = arg.span.into();
                     span.push_span_label(
```

Comment:
also I feel like we should not emit the "consider changing this parameter type in function `generic` to borrow instead if owning the value isn't necessary" note if a reborrow would work. so either suggest a reborrow or suggest changing the fn_sig, but not both

Reply:
> also I feel like we should not emit the "consider changing this parameter type in function `generic` to borrow instead if owning the value isn't necessary" note if a reborrow would work. so either suggest a reborrow or suggest changing the fn_sig, but not both

Done. Thank you.